### PR TITLE
Handle symlinks in LocalFileSystem (#2206)

### LINKED ISSUE #2206: object_store: Create explicit test for symlinks
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

As described on https://github.com/apache/arrow-rs/pull/2195#discussion_r931569201 we are currently relying on idiosynchroncies of the MacOS contents of `/` for coverage of handling broken symlinks. 

**Describe the solution you'd like**
Test that creates symlinks  and ensures listing still works

This probably should include good symlinks as well as broken symlinks

**Describe alternatives you've considered**
N/A

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2206

# Rationale for this change
 

# What changes are included in this PR?

This attempts to handle symlinks consistently

# Are there any user-facing changes?

Yes, symlinks will behave as expected

### CODE REVIEW

File: object_store/src/path/mod.rs [hunk 1]
```diff
@@ -295,20 +294,13 @@ where
     }
 }
 
-/// Given a filesystem path, convert it to its canonical URL representation,
-/// returning an error if the file doesn't exist on the local filesystem
+/// Given a filesystem path convert it to a URL representation
 pub(crate) fn filesystem_path_to_url(
     path: impl AsRef<std::path::Path>,
 ) -> Result<Url, Error> {
-    let path = path.as_ref().canonicalize().context(CanonicalizeSnafu {
```
Comment (tustvold) [inline object_store/src/path/mod.rs:303]: This is the change that makes this possible, we no longer canonicalize the paths

File: object_store/src/local.rs [hunk 2]
```diff
@@ -433,7 +446,10 @@ impl ObjectStore for LocalFileSystem {
         let resolved_prefix = config.path_to_filesystem(&prefix)?;
 
         maybe_spawn_blocking(move || {
-            let walkdir = WalkDir::new(&resolved_prefix).min_depth(1).max_depth(1);
+            let walkdir = WalkDir::new(&resolved_prefix)
+                .min_depth(1)
+                .max_depth(1)
+                .follow_links(true);
```
Comment (tustvold) [inline object_store/src/local.rs:452]: Amusingly the lack of `follow_links` previously meant that symlinks would be reported as `objects` even if they were actually directories :sweat_smile:

File: object_store/src/local.rs [hunk 3]
```diff
@@ -214,10 +225,11 @@ impl LocalFileSystem {
 
 impl Config {
     /// Return filesystem path of the given location
-    fn path_to_filesystem(&self, location: &Path) -> Result<std::path::PathBuf> {
+    fn path_to_filesystem(&self, location: &Path) -> Result<PathBuf> {
         let mut url = self.root.clone();
         url.path_segments_mut()
             .expect("url path")
+            .pop_if_empty()
```
Comment (tustvold) [inline object_store/src/local.rs:232]: Drive by fix, this technically isn't necessary as `Path` ignores empty segments, but this avoids creating paths with "//" which look odd in error messages.

File: object_store/src/local.rs [hunk 4]
```diff
@@ -170,6 +170,17 @@ impl From<Error> for super::Error {
 ///
 /// If not called from a tokio context, this will perform IO on the current thread with
 /// no additional complexity or overheads
+///
+/// # Symlinks
```
Comment (alamb) [inline object_store/src/local.rs:174]: 👍

File: object_store/src/local.rs (hunk 3 above)
Comment (alamb) [reply object_store/src/local.rs:232]:
```suggestion
            // technically not necessary as Path ignores empty segments
            // but avoids creating paths with "//" which look odd in error messages.
            .pop_if_empty()
```

File: object_store/src/local.rs [hunk 5]
```diff
@@ -1030,6 +1047,124 @@ mod tests {
...
+        let file = subdir.join("file.parquet");
+        std::fs::write(file, "test").unwrap();
+
+        check_list(&integration, None, &["a/file.parquet"]).await;
+        integration
+            .head(&Path::from("a/file.parquet"))
+            .await
+            .unwrap();
+
+        // Follow out of tree symlink
```
Comment (alamb) [inline object_store/src/local.rs:1085]: love the comments in this test

**Review by alamb — APPROVED** [review]
This looks awesome -- thank you @tustvold

**Pull request merged**